src/sessions/middleware.rs

```diff
@@ -0,0 +1,255 @@
+    }
+
+    /// Sets a session ttl. This will be used both for the cookie
+    /// expiry and also for the session-internal expiry.
+    ///
+    /// The default for this value is 1 day. Set this to None to not
+    /// set a cookie or session expiry. This is not recommended.
+    pub fn with_session_ttl(mut self, session_ttl: Option<Duration>) -> Self {
+        self.session_ttl = session_ttl;
+        self
+    }
+
+    /// Sets the name of the cookie that the session is stored with or in.
+    ///
+    /// If you are running multiple tide applications on the same
+    /// domain, you will need different values for each
+    /// application. The default value is "tide.sid"
+    pub fn with_cookie_name(mut self, cookie_name: String) -> Self {
+        self.cookie_name = cookie_name;
+        self
+    }
+
+    async fn load_or_create(&self, cookie_value: Option<String>) -> Session {
+        let session = match cookie_value {
+            Some(cookie_value) => self.store.load_session(cookie_value).await,
+            None => None,
+        };
+
+        session.unwrap_or_default()
+    }
+
+    fn build_cookie(&self, secure: bool, cookie_value: String) -> Cookie<'static> {
+        let mut cookie = Cookie::build(self.cookie_name.clone(), cookie_value)
+            .http_only(true)
+            .same_site(SameSite::Strict)
+            .secure(secure)
+            .path(self.cookie_path.clone())
+            .finish();
+
+        if let Some(ttl) = self.session_ttl {
+            cookie.set_expires(Some((std::time::SystemTime::now() + ttl).into()));
+        }
+
+        self.sign_cookie(&mut cookie);
+
+        cookie
+    }
+
+    /// Signs the cookie's value providing integrity and authenticity.
+    fn sign_cookie(&self, cookie: &mut Cookie<'_>) {
+        // Compute HMAC-SHA256 of the cookie's value.
+        let mut mac = Hmac::<Sha256>::new_varkey(&self.key.signing()).expect("good key");
+        mac.update(cookie.value().as_bytes());
+
+        // Cookie's new value is [MAC | original-value].
+        let mut new_value = base64::encode(&mac.finalize().into_bytes());
+        new_value.push_str(cookie.value());
+        cookie.set_value(new_value);
+    }
+
+    /// Given a signed value `str` where the signature is prepended to `value`,
+    /// verifies the signed value and returns it. If there's a problem, returns
+    /// an `Err` with a string describing the issue.
+    fn verify_signature(&self, cookie_value: &str) -> Result<String, &'static str> {
+        if cookie_value.len() < BASE64_DIGEST_LEN {
+            return Err("length of value is <= BASE64_DIGEST_LEN");
+        }
+
+        // Split [MAC | original-value] into its two parts.
+        let (digest_str, value) = cookie_value.split_at(BASE64_DIGEST_LEN);
+        let digest = base64::decode(digest_str).map_err(|_| "bad base64 digest")?;
+
+        // Perform the verification.
+        let mut mac = Hmac::<Sha256>::new_varkey(&self.key.signing()).expect("good key");
```

Comment:
_`Hmac` can't actually fail as of the current time, but maybe it should at some future version? Dunno._
```suggestion
        let mut mac = Hmac::<Sha256>::new_varkey(&self.key.signing()).expect("Hmac key failure");
```

Reply:
this code block is verbatim from Cookie

Reply:
marking similar pr review comments as resolved, since changes to this code should probably be introduced in the cookie crate. keeping this code as similar as possible to the upstream code will make audits and downstreaming patches easier